platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesBirthRoundMigration.java

```diff
@@ -0,0 +1,262 @@
+
+    private PcesBirthRoundMigration() {}
+
+    /**
+     * Migrate a PCES in generation mode to a PCES in birth round mode if needed. No op if the migration has already
+     * been completed.
+     *
+     * @param platformContext                        the platform context
+     * @param migrationRound                         the round at which the migration is occurring, this will be equal
+     *                                               to the round number of the initial state
+     * @param minimumJudgeGenerationInMigrationRound the minimum judge generation in the migration round
+     */
+    public static void migratePcesToBirthRoundMode(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final RecycleBin recycleBin,
+            @NonNull final NodeId selfId,
+            final long migrationRound,
+            final long minimumJudgeGenerationInMigrationRound)
+            throws IOException {
+
+        final Path databaseDirectory = getDatabaseDirectory(platformContext, selfId);
+
+        if (findPcesFiles(databaseDirectory, GENERATION_THRESHOLD).isEmpty()) {
+            // No migration needed if there are no PCES files in generation mode.
+
+            logger.info(STARTUP.getMarker(), "PCES birth round migration is not necessary.");
+            return;
+        } else if (!findPcesFiles(databaseDirectory, BIRTH_ROUND_THRESHOLD).isEmpty()) {
+            // We've found PCES files in both birth round and generation mode.
+            // This is a signal that we attempted to do the migration but crashed.
+            // The migrated PCES file is written atomically, so if it exists,
+            // the important part of the migration has been completed. Remaining
+            // work is to clean up the old files.
+
+            logger.error(
+                    EXCEPTION.getMarker(),
+                    "PCES birth round migration has already been completed, but there "
+                            + "are still legacy formatted PCES files present. Cleaning up.");
+            makeBackupFiles(recycleBin, databaseDirectory);
+            cleanUpOldFiles(databaseDirectory);
+
+            return;
+        }
+
+        logger.info(
+                STARTUP.getMarker(),
+                "Migrating PCES to birth round mode. Migration round: {}, minimum judge "
+                        + "generation in migration round: {}.",
+                migrationRound,
+                minimumJudgeGenerationInMigrationRound);
+
+        makeBackupFiles(recycleBin, databaseDirectory);
+
+        final List<GossipEvent> eventsToMigrate = readEventsToBeMigrated(
+                platformContext, recycleBin, selfId, minimumJudgeGenerationInMigrationRound, migrationRound);
+
+        if (eventsToMigrate.isEmpty()) {
+            logger.error(EXCEPTION.getMarker(), "No events to migrate. PCES birth round migration aborted.");
```

Comment:
Is this case supposed to be impossible (it isn't being tested)? If it isn't impossible, should old files be cleaned up?

Reply:
This should be impossible. More of a sanity check than a real scenario I expect to encounter.